### PR TITLE
Add integration test for npm multiple packages

### PR DESCRIPTION
**Pull request by ejegrova**
Expected behavior for packages: should be handled the same way, should be both installable, the SBOM should report the union of dependencies from both packages

JIRA: STONEBLD-1419

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- n/a Code coverage from testing does not decrease and new code is covered
- n/a Docs updated (if applicable)
- n/a Docs links in the code are still valid (if docs were updated)

### CODE REVIEW

File: tests/integration/test_npm.py [hunk 1]
```diff
@@ -77,6 +77,30 @@
...
+            utils.TestParameters(
+                repo="https://github.com/cachito-testing/npm-cachi2-multiple-packages.git",
+                ref="b37466c6a680bb82e087c17ac0232b554f7f2cc6",
+                packages=(
+                    {"path": "first_pkg", "type": "npm"},
+                    {"path": "second_pkg", "type": "npm"},
+                ),
+                check_vendor_checksums=False,
+            ),
+            id="npm_multiple_packages_lockfile1",
```
Comment (lkolacek) [inline tests/integration/test_npm.py:90]: We might drop support for V1. See my PR https://github.com/containerbuildsystem/cachi2/pull/229

Comment (lkolacek) [reply tests/integration/test_npm.py:90]: Please provide your feedback on dropping V1.

File: tests/integration/test_npm.py [hunk 2]
```diff
@@ -77,6 +77,30 @@
...
+        ),
+        pytest.param(
+            utils.TestParameters(
+                repo="https://github.com/cachito-testing/npm-cachi2-multiple-packages.git",
```
Comment (lkolacek) [inline tests/integration/test_npm.py:94]: Could you pls update the Readme file in the test data repo with the expected behavior info?

**Comment by brunoapimentel** [comment]
A question about the [repo](https://github.com/cachito-testing/npm-cachi2-multiple-packages/tree/main) we're using for these tests: are the folders bar, baz, eggs-packages, etc needed? It seems we're not even referencing them in package.json.

**Comment by ejegrova** [comment]
> A question about the [repo](https://github.com/cachito-testing/npm-cachi2-multiple-packages/tree/main) we're using for these tests: are the folders bar, baz, eggs-packages, etc needed? It seems we're not even referencing them in package.json.

removed those

File: tests/integration/test_npm.py (hunk 2 above)
Comment (ejegrova) [reply tests/integration/test_npm.py]: updated: https://github.com/cachito-testing/npm-cachi2-multiple-packages/blob/lockfile-v1/README.md

**Comment by lkolacek** [comment]
@ejegrova you hit  The 'node_modules' directory cannot be present in the source repository, which Ben merged a while ago. Update source repo and remove pre-fetched data pls.

File: tests/integration/test_data/npm_multiple_packages_lockfile3/fetch_deps_sha256sums.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (chmeliik) [inline tests/integration/test_data/npm_multiple_packages_lockfile3/fetch_deps_sha256sums.json:1]: Something is very wrong here :thinking:

File: tests/integration/test_data/npm_multiple_packages_lockfile3/bom.json [hunk 4]
```diff
@@ -0,0 +1,1431 @@
...
+  "components": [
+    {
+      "name": "accepts",
+      "properties": [
+        {
+          "name": "cachi2:found_by",
+          "value": "cachi2"
+        }
+      ],
+      "purl": "pkg:npm/accepts@1.3.8?vcs_url=git%2Bhttps://github.com/cachito-testing/npm-cachi2-multiple-packages.git%40c19e39bbd6588ee11ab3470a99763fcdbf49d1de#first_pkg/node_modules/accepts",
```
Comment (chmeliik) [inline tests/integration/test_data/npm_multiple_packages_lockfile3/bom.json:12]: And here as well :sweat_smile:

Comment (chmeliik) [reply tests/integration/test_data/npm_multiple_packages_lockfile3/bom.json:12]: Hmm, somehow all the "resolved" urls disappeared in https://github.com/cachito-testing/npm-cachi2-multiple-packages/commit/c19e39bbd6588ee11ab3470a99763fcdbf49d1de, that's probably why

Comment (ejegrova) [reply tests/integration/test_data/npm_multiple_packages_lockfile3/bom.json]: seems like regenerating files helped

File: tests/integration/test_npm.py [hunk 5]
```diff
@@ -30,6 +30,18 @@
...
             id="npm_lockfile3_yarn_registry",
         ),
+        pytest.param(
+            utils.TestParameters(
+                repo="https://github.com/cachito-testing/npm-cachi2-multiple-packages.git",
+                ref="b50c11c6f49f6df8fdfc42e6925d7ac4120c3bdd",
+                packages=(
+                    {"path": "first_pkg", "type": "npm"},
+                    {"path": "second_pkg", "type": "npm"},
+                ),
```
Comment (chmeliik) [inline tests/integration/test_npm.py:40]: The [first_pkg](https://github.com/cachito-testing/npm-cachi2-multiple-packages/blob/b50c11c6f49f6df8fdfc42e6925d7ac4120c3bdd/first_pkg/package.json#L20) and [second_pkg](https://github.com/cachito-testing/npm-cachi2-multiple-packages/blob/b50c11c6f49f6df8fdfc42e6925d7ac4120c3bdd/second_pkg/package.json#L21) have exactly the same dependencies, which means cachi2 could entirely ignore one of them and this test wouldn't notice. Both need some unique dependencies

**Review by chmeliik — COMMENTED** [review]
The repo could use a cleanup 
* the main branch is not being used for testing but has package files
* the lockfile-v1 branch can probably be removed
* maybe lockfile-v3 should be the main branch?

The README mentions that both packages should be installable, but these tests are not e2e. I don't remember why I filed the stories as not-e2e :thinking: 

WDYT, should we fix the tests or the README?

File: tests/integration/test_npm.py (hunk 5 above)
Comment (ejegrova) [reply tests/integration/test_npm.py]: Added more dependencies in packages and the test is e2e now. In npm cache it lists dependencies from both packages.

**Comment by ejegrova** [comment]
> The repo could use a cleanup
> 
>     * the main branch is not being used for testing but has package files
> 
>     * the lockfile-v1 branch can probably be removed
> 
>     * maybe lockfile-v3 should be the main branch?
> 
> 
> The README mentions that both packages should be installable, but these tests are not e2e. I don't remember why I filed the stories as not-e2e thinking
> 
> WDYT, should we fix the tests or the README?

check the repo again please, I left there only one branch which was previously lockfile-v3 branch

File: tests/integration/test_npm.py [hunk 6]
```diff
@@ -262,6 +262,27 @@ def test_npm_smoketest(
...
+            ),
+            ["npm", "cache", "ls"],
+            [
+                "/tmp/npm_multiple_packages_lockfile3-output/deps/npm/types-react-18.2.18.tgz",
+                "/tmp/npm_multiple_packages_lockfile3-output/deps/npm/external-cachito-npm-without-deps/cachito-npm-without-deps-external-sha512-43e71f90ad5f9eb349ab18a283f8954994def373962ddc61b866bdea4d48249e67913c6b84dca1e8c519e981ca1fcc62b438292104a88ee9ed72db76a41efede.tgz",
+                "/tmp/npm_multiple_packages_lockfile3-output/deps/npm/github.com/kevva/is-positive/is-positive-external-gitcommit-97edff6f525f192a3f83cea1944765f769ae2678.tgz",
+                "/tmp/npm_multiple_packages_lockfile3-output/deps/npm/express-4.18.2.tgz",
+                "/tmp/npm_multiple_packages_lockfile3-output/deps/npm/bitbucket.org/cachi-testing/cachi2-without-deps-second/cachi2-without-deps-second-external-gitcommit-09992d418fc44a2895b7a9ff27c4e32d6f74a982.tgz",
+                "/tmp/npm_multiple_packages_lockfile3-output/deps/npm/sax-0.1.1.tgz",
+            ],
```
Comment (chmeliik) [inline tests/integration/test_npm.py:283]: minor: I think we're not checking the `debug` dependency of [second_pkg](https://github.com/cachito-testing/npm-cachi2-multiple-packages/blob/542312f9462dd6549ec57cd95106104c30054ec9/second_pkg/package.json#L21)

**Review by chmeliik — APPROVED** [review]
LGTM with a tiny nitpick :+1: 

Checked that the SBOM reports the union of dependencies (8 unique to first_pkg, 5 unique to second_pkg, 58 common = 71 total) and that we download 69 files (71 - 2 main packages)

File: tests/integration/test_npm.py (hunk 6 above)
Comment (ejegrova) [reply tests/integration/test_npm.py]: added, must have missed that one

**Comment by ejegrova** [comment]
pipeline is failing, fix: https://github.com/containerbuildsystem/cachi2/pull/272

**Comment by lkolacek** [comment]
> LGTM with a tiny nitpick 👍
> 
> Checked that the SBOM reports the union of dependencies (8 unique to first_pkg, 5 unique to second_pkg, 58 common = 71 total) and that we download 69 files (71 - 2 main packages)

LGTM :+1: @chmeliik  should we add a general check to E2e scenarios to test that number of components in SBOM report matches number of downloaded and installed deps?

**Review by lkolacek — APPROVED** [review]

**Comment by chmeliik** [comment]
> > LGTM with a tiny nitpick +1
> > Checked that the SBOM reports the union of dependencies (8 unique to first_pkg, 5 unique to second_pkg, 58 common = 71 total) and that we download 69 files (71 - 2 main packages)
> 
> LGTM +1 @chmeliik should we add a general check to E2e scenarios to test that number of components in SBOM report matches number of downloaded and installed deps?

Could be nice for all scenarios, not just e2e. But we would have to exclude local dependencies and [gomod](https://github.com/containerbuildsystem/cachi2/blob/main/tests/integration/test_data/gomod_e2e_multiple_modules/fetch_deps_sha256sums.json) would need some extra logic

**Comment by chmeliik** [comment]
> Could be nice for all scenarios, not just e2e. But we would have to exclude local dependencies and [gomod](https://github.com/containerbuildsystem/cachi2/blob/main/tests/integration/test_data/gomod_e2e_multiple_modules/fetch_deps_sha256sums.json) would need some extra logic

And for gomod we probably can't actually tell which dependencies are local from the SBOM

**Pull request merged**